kernel/sched.c

```diff
@@ -492,63 +507,33 @@ void k_thread_time_slice_set(struct k_thread *th, int32_t slice_ticks,
 }
 #endif
 
-static inline bool sliceable(struct k_thread *thread)
-{
-	bool ret = is_preempt(thread)
-		&& !z_is_thread_prevented_from_running(thread)
-		&& !z_is_prio_higher(thread->base.prio, slice_max_prio)
-		&& !z_is_idle_thread_object(thread);
-
-#ifdef CONFIG_TIMESLICE_PER_THREAD
-	ret |= thread->base.slice_ticks != 0;
-#endif
-
-	return ret;
-}
-
-static k_spinlock_key_t slice_expired_locked(k_spinlock_key_t sched_lock_key)
-{
-	struct k_thread *curr = _current;
-
-#ifdef CONFIG_TIMESLICE_PER_THREAD
-	if (curr->base.slice_expired) {
-		k_spin_unlock(&sched_spinlock, sched_lock_key);
-		curr->base.slice_expired(curr, curr->base.slice_data);
-		sched_lock_key = k_spin_lock(&sched_spinlock);
-	}
-#endif
-	if (!z_is_thread_prevented_from_running(curr)) {
-		move_thread_to_end_of_prio_q(curr);
-	}
-	z_reset_time_slice(curr);
-
-	return sched_lock_key;
-}
-
 /* Called out of each timer interrupt */
 void z_time_slice(void)
 {
 	k_spinlock_key_t key = k_spin_lock(&sched_spinlock);
+	struct k_thread *curr = _current;
 
 #ifdef CONFIG_SWAP_NONATOMIC
-	if (pending_current == _current) {
-		z_reset_time_slice(_current);
+	if (pending_current == curr) {
+		z_reset_time_slice(curr);
 		k_spin_unlock(&sched_spinlock, key);
 		return;
 	}
 	pending_current = NULL;
 #endif
 
-	if (slice_time(_current) && sliceable(_current)) {
-		if (slice_expired[_current_cpu->id]) {
-			/* Note: this will (if so enabled) internally
-			 * drop and reacquire the scheduler lock
-			 * around the callback!  Don't put anything
-			 * after this line that requires
-			 * synchronization.
-			 */
-			key = slice_expired_locked(key);
+	if (sliceable(curr) && slice_expired[_current_cpu->id]) {
```

Comment:
This bit seems like a mild performance regression for the (common) non-timeslicing case: the previous version would short circuit rapidly if the current thread was not configured to slice (which is just checking a global variable for zero, also the thread struct equivalent if PER_THREAD).  Now it needs to go down the list of conditions in sliceable(), which is a more involved test.